wholegames-httptest2.Rmd

```diff
@@ -0,0 +1,161 @@
+  test_that("gh_organizations errors if the API doesn't behave", {
+    testthat::expect_snapshot_error(gh_organizations())
+  })
+},
+simplify = FALSE)
+```
+
+The first test is similar to what we did for `gh_api_status()` except we didn't touch the mock file this time, out of laziness.
+In the second test there is more to unpack: how do we get a mock file corresponding to an error?
+
+* We first run the test as is. It fails because there is no error, which we expected. Note the `simplify = FALSE` that means the mock file also contains headers for the response.
+* We replaced `200L` with `502L` and removed the body, to end up with a simpler mock file under ` tests/testthat/gh_organizations_error/api.github.com/organizations-5377e8.R`
+
+```r
+structure(list(method = "GET", url = "https://api.github.com/organizations?since=1",
+    status_code = 502L, headers = structure(list(server = "GitHub.com",
+        date = "Thu, 17 Feb 2022 12:40:29 GMT", `content-type` = "application/json; charset=utf-8",
+        `cache-control` = "private, max-age=60, s-maxage=60",
+        vary = "Accept, Authorization, Cookie, X-GitHub-OTP",
+        etag = "W/\"d56e867402a909d66653b6cb53d83286ba9a16eef993dc8f3cb64c43b66389f4\"",
+        `x-oauth-scopes` = "gist, repo, user, workflow", `x-accepted-oauth-scopes` = "",
+        `x-github-media-type` = "github.v3; format=json", link = "<https://api.github.com/organizations?since=3428>; rel=\"next\", <https://api.github.com/organizations{?since}>; rel=\"first\"",
+        `x-ratelimit-limit` = "5000", `x-ratelimit-remaining` = "4986",
+        `x-ratelimit-reset` = "1645104327", `x-ratelimit-used` = "14",
+        `x-ratelimit-resource` = "core", `access-control-expose-headers` = "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
+        `access-control-allow-origin` = "*", `strict-transport-security` = "max-age=31536000; includeSubdomains; preload",
+        `x-frame-options` = "deny", `x-content-type-options` = "nosniff",
+        `x-xss-protection` = "0", `referrer-policy` = "origin-when-cross-origin, strict-origin-when-cross-origin",
+        `content-security-policy` = "default-src 'none'", vary = "Accept-Encoding, Accept, X-Requested-With",
+        `content-encoding` = "gzip", `x-github-request-id` = "A4BA:12D5C:178438:211160:620E423C"), class = "httr2_headers"),
+    body = charToRaw("")), class = "httr2_response")
+
+```
+* We re-run the tests. We got the expected error message.
+
+Without the HTTP testing infrastructure, testing for behavior of the package in case of API errors would be more difficult.
+
+Regarding our secret API token, since httptest2 doesn't save the requests, and since the responses don't contain the token, it is safe without our making any effort.
```

Comment:
Actually, `httr2_response` objects, unlike the equivalent in `httr`, don't include the request. So httptest2 doesn't need to do anything special to drop the request, it isn't there to begin with (one of the many things I could delete when moving to httr2).